test/tools/ossfuzz/SolidityGenerator.h

```diff
@@ -22,21 +22,188 @@
 
 #pragma once
 
+#include <test/tools/ossfuzz/Generators.hpp>
+
+#include <liblangutil/Exceptions.h>
+
+#include <memory>
 #include <random>
+#include <set>
+#include <variant>
 
 namespace solidity::test::fuzzer
 {
+/// Forward declarations
+class SolidityGenerator;
+
+/// Type declarations
+#define SEMICOLON() ;
+#define FORWARDDECLAREGENERATORS(G) class G
+GENERATORLIST(FORWARDDECLAREGENERATORS, SEMICOLON(), SEMICOLON())
+#undef FORWARDDECLAREGENERATORS
+#undef SEMICOLON
+
+#define COMMA() ,
+using GeneratorPtr = std::variant<
+#define VARIANTOFSHARED(G) std::shared_ptr<G>
+GENERATORLIST(VARIANTOFSHARED, COMMA(), )
+>;
+#undef VARIANTOFSHARED
+using Generator = std::variant<
+#define VARIANTOFGENERATOR(G) G
+GENERATORLIST(VARIANTOFGENERATOR, COMMA(), )
+>;
+#undef VARIANTOFGENERATOR
+#undef COMMA
 using RandomEngine = std::mt19937_64;
+using Distribution = std::uniform_int_distribution<size_t>;
+
+struct GenerationProbability
+{
+	static size_t distributionOneToN(size_t _n, std::shared_ptr<RandomEngine> _rand)
+	{
+		return Distribution(1, _n)(*_rand);
+	}
+};
+
+struct AddDependenciesVisitor
+{
+	template <typename T>
+	void operator()(T const& _t)
+	{
+		_t->setup();
+	}
+};
 
-class SolidityGenerator
+struct GeneratorVisitor
+{
+	template <typename T>
+	std::string operator()(T const& _t)
+	{
+		return _t->visit();
+	}
+};
+
+struct ResetVisitor
+{
+	template <typename T>
+	void operator()(T const& _t)
+	{
+		_t->reset();
+	}
+};
+
+struct GeneratorBase
+{
+	GeneratorBase(std::shared_ptr<SolidityGenerator> _mutator);
+	template <typename T>
+	std::shared_ptr<T> generator()
+	{
+		for (auto &g: generators)
+			if (std::holds_alternative<std::shared_ptr<T>>(g))
+				return std::get<std::shared_ptr<T>>(g);
+		solAssert(false, "");
+	}
+	/// Generator
+	virtual std::string visit() = 0;
+	std::string visitChildren();
+	void addGenerators(std::set<GeneratorPtr> _generators)
+	{
+		for (auto &g: _generators)
+			generators.insert(g);
```

Comment:
I don't know if this code links `solutil`, but if it does you could just do `generators += _generators;` I think